### ISSUE TITLE
DP-FedAvg and DP-FedSGD

### ISSUE THREAD

**Issue by dionman** [tensorflow/federated#745]
Is there an implementation of DP-FedAvg and DP-FedSGD available/to be added?

**Comment by kairouzp** [comment]
TF Privacy is designed to be compatible with TFF.  Integrating these two should be easy, and we are going to do it very soon.

**Comment by klize** [comment]
@kairouzp 
Then is it impossible to use optimizers such as DPGradientDescentGaussianOptimizer() with tf.keras yet?

Actually I tried to do this thing like below but it falis with the error; 'TFOptimizer' object has no attribute 'variables'
```python
def model_fn():
  model = tf.keras.models.Sequential([ ... ])
  model.compile(optimizer=DPGradientDescentGaussianOptimizer(..), ... )

  return tff.learning.from_compiled_keras_model(model,batch)
```

**Comment by g0wzancha** [comment]
Hi, I have a follow-up question.
1. As far as I understood, TF Privacy clips mini-gradients (derived from mini-batches), averages them, then adds the noise. The whole process is on one device.
2. For federated learning, McMahan et al. [1] suggests clipping mini-gradients on the user device, then on the server side averaging them and adding the noise.

Just to be sure, the code [2] implements the latter case (averaging and adding noise on the server)?
Would it be possible to adjust the arguments of "tff.learning.federated_averaging.build_federated_averaging_process( )", in order to implement the first case?

[1] https://arxiv.org/pdf/1710.06963.pdf
[2] https://github.com/tensorflow/federated/blob/master/tensorflow_federated/python/research/differential_privacy/stackoverflow/run_federated.py

**Comment by ZacharyGarrett** [comment]
[`tff.utils.build_dp_aggregate`](https://www.tensorflow.org/federated/api_docs/python/tff/utils/build_dp_aggregate`) provides an API for combining TF privacy queries with TFF. The object it returns is intended to be given to [`tff.learning.build_federated_averaging_process`](https://www.tensorflow.org/federated/api_docs/python/tff/learning/build_federated_averaging_process) via the `stateful_delta_aggregate_fn` parameter.

**Comment by kairouzp** [comment]
You are right that [2] implements the latter (i.e. bullet 2. in your
email). This is needed to train "federated models" with "user-level"
differential privacy (DP).

You can certainly use the combination of TFF + TF Privacy to train with
user-level DP even when the data is all on one device. This is doable as
long as you have the data keyed by clients/users. But if you do not care
about "user-level" DP then you might as well just use TF Privacy (say for
record-level DP).

Does that make sense?

**Comment by g0wzancha** [comment]
@ZacharyGarrett thanks! yes, I am aware of that.
@kairouzp, thanks! yes, I would like to use TF Privacy for record-level DP on the user side in federated learning, and I am not sure if that is possible using the current TFF APIs, because I think it requires to make changes on [1], more precisely, moving the call of `query.get_noised_result( )` from `post_process( )` to `preprocess_record( )`, meaning to add the noise on the user side before collecting and averaging the gradients by the server. Does it make sense? or is there any other alternative to handle this?

[1] https://github.com/tensorflow/federated/blob/80c068e6e644b8cf2e9024909671213737d60c3a/tensorflow_federated/python/core/utils/differential_privacy.py

**Comment by kairouzp** [comment]
+Galen Andrew <galenandrew@google.com> is a better person to answer this
question, but my feel is that you can do that.

Having said that, I am not sure why you would want to train with on-device
record-level DP? What is the meaning of that?

**Comment by kairouzp** [comment]
Yes you could do what you want with fairly minimal changes. You wouldn't
actually change TFF, it's TFP you'd need to change (or rather extend). You
would have to make a new DPQuery in TFP that is almost the same as
GaussianSumQuery
<https://github.com/tensorflow/privacy/blob/master/tensorflow_privacy/privacy/dp_query/gaussian_query.py>,
except the addition of noise would move from get_noised_result to
preprocess_record. (get_noised_result would simply return the sample_state
and global_state unchanged.) Then make one of your new DPQueries and pass
it to build_dp_aggregate in TFF.

**Comment by g0wzancha** [comment]
Thank you very much Peter for the hints, quite useful!

> Having said that, I am not sure why you would want to train with on-device
> record-level DP? What is the meaning of that?

Actually I have no clear hypothesis at this point, however, in my research in general, I attempt to  push the implementation of privacy practices more to the user side. For now, just trying out things, let's see if that will lead me somewhere :)

**Comment by cvarun16** [comment]
Following up with this, is there a working example of (a) central DP with tff, and (b) local DP with tff in the repository?

**Comment by kairouzp** [comment]
There are working examples for FL + central DP in TFF:
https://github.com/tensorflow/federated/blob/master/tensorflow_federated/python/research/differential_privacy/emnist/run_federated.py
(for
EMNIST) and
https://github.com/tensorflow/federated/blob/master/tensorflow_federated/python/research/differential_privacy/stackoverflow/run_federated.py
(for
Stack Overflow). We do not have an example for FL + local DP in TFF, but it
wouldn't be too hard to do it. What you would need to do is modify the
GaussianSumQuery (
https://github.com/tensorflow/privacy/blob/master/tensorflow_privacy/privacy/dp_query/gaussian_query.py)
to make it add noise to each record rather than the sum. In particular, you
want to move the noise addition from this line
<https://github.com/tensorflow/privacy/blob/master/tensorflow_privacy/privacy/dp_query/gaussian_query.py#L99>
to
right after this line
<https://github.com/tensorflow/privacy/blob/master/tensorflow_privacy/privacy/dp_query/gaussian_query.py#L81>.
You can change the l2 clipping step
<https://github.com/tensorflow/privacy/blob/master/tensorflow_privacy/privacy/dp_query/gaussian_query.py#L81>
to
l1 clipping <https://stanford.edu/~jduchi/projects/DuchiShSiCh08.pdf>
(which is a little tricky in TF) and the noise from Gaussian to Laplace if
you want to get pure local DP.  Of course, I recommend you also edit the
name of the query (perhaps to LocalGaussianSumQuery if you choose to keep
the Gaussian noise and to LocalLaplaceSumQuery if you choose to use Laplace
noise). After this step, you would need to change the `build_dp_query
<https://github.com/tensorflow/federated/blob/07ec93478c4e88fcd8cfa4afad2ca6145bfb091d/tensorflow_federated/python/core/utils/differential_privacy.py#L39>`
to make it take the new query you have written (say LocalGaussianSumQuery
instead of GaussianSumQuery). Hope this helps!

**Comment by cvarun16** [comment]
Thanks Peter; will reach out if there are any issues. Appreciate your detailed post.

**Comment by cvarun16** [comment]
Thanks again Peter; this is quite straightforward. I am assuming these modifications impact privacy expenditure calculation defined in https://github.com/tensorflow/privacy/blob/master/tensorflow_privacy/privacy/analysis/compute_dp_sgd_privacy_lib.py#L57, which calls a method in https://github.com/tensorflow/privacy/blob/c948e2fe7c85ea5f73bce3951b3876e8d5a39ce0/tensorflow_privacy/privacy/analysis/rdp_accountant.py#L267 ?

**Comment by kairouzp** [comment]
Good question. It depends on what privacy budget you want to track. If you
use the Gaussian mechanism locally to achieve (eps, delta) local DP, you
would get (eps/sqrt(n), delta) central DP. This is because the sum of
Gaussians is a Gaussian with sqrt(n)*sigma (hence the reduction by a factor
of sqrt(n) epsilon). So if you want to track the central DP privacy budget,
all you have to do is modify the part where you write the standard
deviation of Gaussian noise to the ledger to make it sqrt(n)*sigma -- and
the rest would still work as is. If you want to track local DP budgets per
user, then you would need to change the entire ledger system to track how
many times you touch each user and apply accounting at the user level. This
is definitely more involved but you could also just report the "per
participation (eps, delta) local DP" and then also report the expected
number of times you expect a user to participate. That should be enough,
especially if you do what I recommended above to track the central DP
budget. Hope this helps!

**Comment by wangdan269** [comment]
> There are working examples for FL + central DP in TFF: https://github.com/tensorflow/federated/blob/master/tensorflow_federated/python/research/differential_privacy/emnist/run_federated.py (for EMNIST) and https://github.com/tensorflow/federated/blob/master/tensorflow_federated/python/research/differential_privacy/stackoverflow/run_federated.py (for Stack Overflow).

The pages of the two examples are 404. Can anyone who downloaded the codes share them with me?
